src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs

```diff
@@ -0,0 +1,129 @@
+            // create the ones we need.  Note that FileExists may fail due
+            // to Win32 ACL's preventing us from seeing a directory, and this
+            // isn't threadsafe.
+
+            bool somepathexists = false;
+
+            int length = fullPath.Length;
+
+            // We need to trim the trailing slash or the code will try to create 2 directories of the same name.
+            if (length >= 2 && Path.EndsInDirectorySeparator(fullPath.AsSpan()))
+                length--;
+
+            int lengthRoot = PathInternal.GetRootLength(fullPath.AsSpan());
+
+            if (length > lengthRoot)
+            {
+                // Special case root (fullpath = X:\\)
+                int i = length - 1;
+                while (i >= lengthRoot && !somepathexists)
+                {
+                    string dir = fullPath.Substring(0, i + 1);
+
+                    if (!DirectoryExists(dir)) // Create only the ones missing
+                        stackDir.Add(dir);
+                    else
+                        somepathexists = true;
+
+                    while (i > lengthRoot && !PathInternal.IsDirectorySeparator(fullPath[i])) i--;
+                    i--;
+                }
+            }
+
+            int count = stackDir.Count;
+            bool r = true;
+            int firstError = 0;
+            string errorString = fullPath;
+
+            fixed (byte* pSecurityDescriptor = securityDescriptor)
+            {
+                Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = new Interop.Kernel32.SECURITY_ATTRIBUTES
+                {
+                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
+
+                while (stackDir.Count > 0)
+                {
+                    string name = stackDir[stackDir.Count - 1];
+                    stackDir.RemoveAt(stackDir.Count - 1);
+
+                    r = Interop.Kernel32.CreateDirectory(name, ref secAttrs);
+                    if (!r && (firstError == 0))
+                    {
+                        int currentError = Marshal.GetLastWin32Error();
+                        // While we tried to avoid creating directories that don't
+                        // exist above, there are at least two cases that will
+                        // cause us to see ERROR_ALREADY_EXISTS here.  FileExists
+                        // can fail because we didn't have permission to the
+                        // directory.  Secondly, another thread or process could
+                        // create the directory between the time we check and the
+                        // time we try using the directory.  Thirdly, it could
+                        // fail because the target does exist, but is a file.
+                        if (currentError != Interop.Errors.ERROR_ALREADY_EXISTS)
+                            firstError = currentError;
```

Comment:
Nit: as long as you're editing this code, please add braces around the body of this if block

Reply:
Sure no problem, I can add the missing braces.